File: components/blitz/src/omero/sys/ParametersI.java
/*
 *   $Id$
 *
 *   Copyright 2008 Glencoe Software, Inc. All rights reserved.
 *   Use is subject to license terms supplied in LICENSE.txt
 *
 */

package omero.sys;

import static omero.rtypes.rbool;
import static omero.rtypes.rint;
import static omero.rtypes.rlist;
import static omero.rtypes.rlong;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import omero.RInt;
import omero.RList;
import omero.RLong;
import omero.RTime;
import omero.RType;

/**
 * Helper subclass of {@link omero.sys.Parameters} for simplifying method
 * parameter creation.
 * 
 * As of 4.0, this takes over for PojoOptions. See ticket:67.
 */
public class ParametersI extends omero.sys.Parameters {

    /**
     * Default constructor creates the {@link #map} instance to prevent later
     * {@link NullPointerException}s. To save memory, it is possible to pass
     * null to {@link ParametersI#ParametersI(Map)}.
     */
    public ParametersI() {
        this.map = new HashMap<String, RType>();
    }

    /**
     * Uses (and does not copy) the given {@code Map<String, RType>} as the
     * named parameter store in this instance. Be careful if either null is
     * passed or if this instance is being used in a multi-threaded environment.
     * No synchronization takes place.
     * @param map the named parameter store to use
     */
    public ParametersI(Map<String, RType> map) {
        this.map = map;
    }

    // ~ Parameters.theFilter.limit & offset
    // =========================================================================

    /**
     * Nulls both the {@link Filter#limit} and {@link Filter#offset} values.
     * @return this instance, for method chaining
     */
    public Parameters noPage() {
        if (this.theFilter != null) {
            this.theFilter.limit = null;
            this.theFilter.offset = null;
        }
        return this;
    }

    /**
     * Sets both the {@link Filter#limit} and {@link Filter#offset} values by
     * wrapping the arguments in {@link RInt} and passing the values to
     * {@link #page(RInt, RInt)}
     * @param offset the offset (to start from)
     * @param limit the limit (maximum to return)
     * @return this instance, for method chaining
     */
    public ParametersI page(int offset, int limit) {
        return this.page(rint(offset), rint(limit));
    }

    /**
     * Creates a {@link Filter} if necessary and sets both {@link Filter#limit}
     * and {@link Filter#offset}.
     * @param offset the offset (to start from)
     * @param limit the limit (maximum to return)
     * @return this instance, for method chaining
     */
    public ParametersI page(RInt offset, RInt limit) {
        if (this.theFilter == null) {
            this.theFilter = new Filter();
        }
        this.theFilter.limit = limit;
        this.theFilter.offset = offset;
        return this;
    }

    /**
     * Returns <code>true</code> if the filter contains a <code>limit</code>
     * <em>OR</em> a <code>offset</code>, <code>false</code> otherwise.
     * 
     * @return See above.
     */
    public boolean isPagination() {
        if (this.theFilter != null) {
            return (null != this.theFilter.limit)
                    || (null != this.theFilter.offset);
        }
        return false;
    }

    /**
     * Returns the value of the <code>offset</code> parameter.
     * 
     * @return See above.
     */
    public RInt getOffset() {
        if (this.theFilter != null) {
            return this.theFilter.offset;
        }
        return null;
    }

    /**
     * Returns the value of the <code>limit</code> parameter.
     * 
     * @return See above.
     */
    public RInt getLimit() {
        if (this.theFilter != null) {
            return this.theFilter.limit;
        }
        return null;
    }
    
    public ParametersI unique() {
        if (this.theFilter == null) {
            this.theFilter = new Filter();
        }
        this.theFilter.unique = rbool(true);
        return this;
    }
    
    public ParametersI noUnique() {
        if (this.theFilter == null) {
            this.theFilter = new Filter();
        }
        this.theFilter.unique = rbool(false);
        return this;
    }
    
    public omero.RBool getUnique() {
        if (this.theFilter != null) {
            return this.theFilter.unique;
        }
        return null;
    }

    // ~ Parameters.theFilter.ownerId & groupId
    // =========================================================================

    /**
     * Sets the value of the <code>experimenter</code> parameter.
     * 
     * @param i
     *            The Id of the experimenter.
     * @return Returns the current object.
     */
    public ParametersI exp(RLong i) {
        if (this.theFilter == null) {
            this.theFilter = new Filter();
        }
        this.theFilter.ownerId = i;
        return this;
    }

    /**
     * Removes the <code>experimenter</code> parameter from the map.
     * 
     * @return Returns the current object.
     */
    public ParametersI allExps() {
        if (this.theFilter != null) {
            this.theFilter.ownerId = null;
        }
        return this;
    }

    /**
     * Returns <code>true</code> if the filter contains and <code>ownerId</code>
     * parameter, <code>false</code> otherwise.
     * 
     * @return See above.
     */
    public boolean isExperimenter() {
        if (this.theFilter != null) {
            return null != this.theFilter.ownerId;
        }
        return false;
    }

    /**
     * Returns the value of the <code>experimenter</code> parameter.
     * 
     * @return See above.
     */
    public RLong getExperimenter() {
        if (this.theFilter != null) {
            return this.theFilter.ownerId;
        }
        return null;
    }

    /**
     * Sets the value of the <code>group</code> parameter.
     * 
     * @param i
     *            The value to set.
     * @return See above.
     */
    public ParametersI grp(RLong i) {
        if (this.theFilter == null) {
            this.theFilter = new Filter();
        }
        this.theFilter.groupId = i;
        return this;
    }

    /**
     * Removes the <code>group</code> parameter from the map.
     * 
     * @return Returns the current object.
     */
    public ParametersI allGrps() {
        if (this.theFilter != null) {
            this.theFilter.groupId = null;
        }
        return this;
    }

    /**
     * Returns <code>true</code> if the filter contains an <code>groupId</code>,
     * <code>false</code> otherwise.
     * 
     * @return See above.
     */
    public boolean isGroup() {
        if (this.theFilter != null) {
            return null != this.theFilter.groupId;
        }
        return false;
    }

    /**
     * Returns the value of the <code>group</code> parameter.
     * 
     * @return See above.
     */
    public RLong getGroup() {
        if (this.theFilter != null) {
            return this.theFilter.groupId;
        }
        return null;
    }

    // ~ Parameters.theFilter.startTime, endTime
    // =========================================================================

    /**
     * Sets the value of the <code>start time</code> parameter.
     * 
     * @param startTime
     *            The time to set.
     * @return Returns the current object.
     */
    public ParametersI startTime(RTime startTime) {
        if (this.theFilter == null) {
            this.theFilter = new Filter();
        }
        this.theFilter.startTime = startTime;
        return this;
    }

    /**
     * Sets the value of the <code>end time</code> parameter.
     * 
     * @param endTime
     *            The time to set.
     * @return Returns the current object.
     */
    public ParametersI endTime(RTime endTime) {
        if (this.theFilter == null) {
            this.theFilter = new Filter();
        }
        this.theFilter.endTime = endTime;
        return this;
    }

    /**
     * Removes the time parameters from the map.
     * 
     * @return Returns the current object.
     */
    public ParametersI allTimes() {
        if (this.theFilter != null) {
            this.theFilter.startTime = null;
            this.theFilter.endTime = null;
        }
        return this;
    }

    /**
     * Returns <code>true</code> if the map contains the <code>start time</code>
     * parameter, <code>false</code> otherwise.
     * 
     * @return See above.
     */
    public boolean isStartTime() {
        if (this.theFilter != null) {
            return null != this.theFilter.startTime;
        }
        return false;
    }

    /**
     * Returns <code>true</code> if the map contains the <code>end time</code>
     * parameter, <code>false</code> otherwise.
     * 
     * @return See above.
     */
    public boolean isEndTime() {
        if (this.theFilter != null) {
            return null != this.theFilter.endTime;
        }
        return false;
    }

    /**
     * Returns the value of the <code>start time</code> parameter.
     * 
     * @return See above.
     */
    public RTime getStartTime() {
        if (this.theFilter != null) {
            return this.theFilter.startTime;
        }
        return null;
    }

    /**
     * Returns the value of the <code>end time</code> parameter.
     * 
     * @return See above.
     */
    public RTime getEndTime() {
        if (this.theFilter != null) {
            return this.theFilter.endTime;
        }
        return null;
    }

    // ~ Parameters.theOption.leaves, orphan, acquisitionData
    // =========================================================================

    /**
     * Sets the <code>leaf</code> parameter to <code>true</code>.
     * 
     * @return Returns the current object.
     */
    public ParametersI leaves() {
        if (this.theOptions == null) {
            this.theOptions = new Options();
        }
        this.theOptions.leaves = rbool(true);
        return this;
    }

    /**
     * Sets the <code>leaf</code> parameter to <code>false</code>.
     * 
     * @return Returns the current object.
     */
    public ParametersI noLeaves() {
        if (this.theOptions == null) {
            this.theOptions = new Options();
        }
        this.theOptions.leaves = rbool(false);
        return this;
    }

    public omero.RBool getLeaves() {
        if (this.theOptions != null) {
            return this.theOptions.leaves;
        }
        return null;
    }

    /**
     * Sets the <code>orphan</code> parameter to <code>true</code>.
     * 
     * @return Returns the current object.
     */
    public ParametersI orphan() {
        if (this.theOptions == null) {
            this.theOptions = new Options();
        }
        this.theOptions.orphan = rbool(true);
        return this;
    }

    /**
     * Sets the <code>orphan</code> parameter to <code>false</code>.
     * 
     * @return Returns the current object.
     */
    public ParametersI noOrphan() {
        if (this.theOptions == null) {
            this.theOptions = new Options();
        }
        this.theOptions.orphan = rbool(false);
        return this;
    }

    public omero.RBool getOrphan() {
        if (this.theOptions != null) {
            return this.theOptions.orphan;
        }
        return null;
    }

    /**
     * Sets the <code>acquisition data</code> parameter to <code>true</code>.
     * 
     * @return Returns the current object.
     */
    public ParametersI acquisitionData() {
        if (this.theOptions == null) {
            this.theOptions = new Options();
        }
        this.theOptions.acquisitionData = rbool(true);
        return this;
    }

    /**
     * Sets the <code>acquisition data</code> parameter to <code>false</code>.
     * 
     * @return Returns the current object.
     */
    public ParametersI noAcquisitionData() {
        if (this.theOptions == null) {
            this.theOptions = new Options();
        }
        this.theOptions.acquisitionData = rbool(false);
        return this;
    }

    public omero.RBool getAcquisitionData() {
        if (this.theOptions != null) {
            return this.theOptions.acquisitionData;
        }
        return null;
    }

    // ~ Parameters.map
    // =========================================================================

    public ParametersI add(String name, RType r) {
        this.map.put(name, r);
        return this;
    }

    public ParametersI addId(long id) {
        add(ome.parameters.Parameters.ID, rlong(id));
        return this;
    }

    public ParametersI addId(RLong id) {
        add(ome.parameters.Parameters.ID, id);
        return this;
    }

    public ParametersI addIds(Collection<Long> longs) {
        addLongs(ome.parameters.Parameters.IDS, longs);
        return this;
    }

    public ParametersI addLong(String name, long l) {
        add(name, rlong(l));
        return this;
    }

    public ParametersI addLong(String name, RLong l) {
        add(name, l);
        return this;
    }

    public ParametersI addLongs(String name, Collection<Long> longs) {
        RList rlongs = rlist();
        for (Long l : longs) {
            rlongs.add(rlong(l));
        }
        this.map.put(name, rlongs);
        return this;
    }

    // ~ Deprecated
    // =========================================================================

    /**
     * Pre-4.0, pojoOptions.map() was a common idiom for passing the {@link Map
     * <String, RType} into methods, to keep those uses valid, the
     * {@link #map()} method is defined.
     * @deprecated use {@link #map()} instead, to be removed in 5.3
     */
    @Deprecated
    public ParametersI map() {
        return this;
    }

}

File: components/tools/OmeroCpp/src/omero/sys/ParametersI.cpp
/*
 *   $Id$
 *
 *   Copyright 2008 Glencoe Software, Inc. All rights reserved.
 *   Use is subject to license terms supplied in LICENSE.txt
 *
 */

#include <omero/sys/ParametersI.h>
#include <omero/RTypesI.h>

using omero::rtypes::rlist;
using omero::rtypes::rlong;
using omero::rtypes::rint;
using omero::rtypes::rbool;

::Ice::Object* IceInternal::upCast(::omero::sys::ParametersI* p) { return p; }

namespace omero {

    namespace sys {

        ParametersI::~ParametersI() {}

        ParametersI::ParametersI(const omero::sys::ParamMap& map) : Parameters() {
	    this->map = map;
        }

	// Parameters.theFilter.limit & offset
	// ===============================================================

	ParametersIPtr ParametersI::noPage() {
	    if (0 != this->theFilter) {
		this->theFilter->limit = omero::RIntPtr();
		this->theFilter->offset = omero::RIntPtr();
	    }
	    return this;
	}

	ParametersIPtr ParametersI::page(Ice::Int offset,
					 Ice::Int limit) {
	    return this->page(rint(offset), rint(limit));
	}

	ParametersIPtr ParametersI::page(const omero::RIntPtr& offset,
					 const omero::RIntPtr& limit) {
	    if (0 == this->theFilter) {
		this->theFilter = new omero::sys::Filter();
	    }
	    this->theFilter->offset = offset;
	    this->theFilter->limit = limit;
	    return this;
	}

	bool ParametersI::isPagination() {
	    if (0 != this->theFilter) {
		return this->theFilter->offset && this->theFilter->limit;
	    }
	    return false;
	}

	omero::RIntPtr ParametersI::getOffset() {
	    if (0 != this->theFilter) {
		return this->theFilter->offset;
	    }
	    return omero::RIntPtr();
	}

	omero::RIntPtr ParametersI::getLimit() {
	    if (0 != this->theFilter) {
		return this->theFilter->limit;
	    }
	    return omero::RIntPtr();
	}

	ParametersIPtr ParametersI::unique() {
	    if (0 == this->theFilter) {
		this->theFilter = new omero::sys::Filter();
	    }
	    this->theFilter->unique = rbool(true);
	    return this;
	}

	ParametersIPtr ParametersI::noUnique() {
	    if (0 == this->theFilter) {
		this->theFilter = new omero::sys::Filter();
	    }
	    this->theFilter->unique = rbool(false);
	    return this;
	}

	omero::RBoolPtr ParametersI::getUnique() {
	    if (0 != this->theFilter) {
		return this->theFilter->unique;
	    }
	    return omero::RBoolPtr();
	}

	// Parameters.theFilter.ownerId & groupId
	// ===============================================================

	ParametersIPtr ParametersI::exp(Ice::Long id) {
	    if (0 == this->theFilter) {
		this->theFilter = new omero::sys::Filter();
	    }
	    this->theFilter->ownerId = rlong(id);
	    return this;
	}

	ParametersIPtr ParametersI::allExps() {
	    if (0 != this->theFilter) {
		this->theFilter->ownerId = omero::RLongPtr();
	    }
	    return this;
	}

	bool ParametersI::isExperimenter() {
	    if (0 != this->theFilter) {
		return this->theFilter->ownerId != 0;
	    }
	    return false;
	}

	omero::RLongPtr ParametersI::getExperimenter() {
	    if (0 != this->theFilter) {
		return this->theFilter->ownerId;
	    }
	    return omero::RLongPtr();
	}

	ParametersIPtr ParametersI::grp(Ice::Long id) {
	    if (0 == this->theFilter) {
		this->theFilter = new omero::sys::Filter();
	    }
	    this->theFilter->groupId = rlong(id);
	    return this;
	}

	ParametersIPtr ParametersI::allGrps() {
	    if (0 != this->theFilter) {
		this->theFilter->groupId = omero::RLongPtr();
	    }
	    return this;
	}

	bool ParametersI::isGroup() {
	    if (0 != this->theFilter) {
		return this->theFilter->groupId != 0;
	    }
	    return false;
	}

	omero::RLongPtr ParametersI::getGroup() {
	    if (0 != this->theFilter) {
		return this->theFilter->groupId;
	    }
	    return omero::RLongPtr();
	}

	// Parameters.theFilter.starttime, endTime
	// ===============================================================

	ParametersIPtr ParametersI::startTime(const omero::RTimePtr& time) {
	    if (0 == this->theFilter) {
		this->theFilter = new omero::sys::Filter();
	    }
	    this->theFilter->startTime = time;
	    return this;
	}

	ParametersIPtr ParametersI::endTime(const omero::RTimePtr& time) {
	    if (0 == this->theFilter) {
		this->theFilter = new omero::sys::Filter();
	    }
	    this->theFilter->endTime = time;
	    return this;
	}

	ParametersIPtr ParametersI::allTimes() {
	    if (0 != this->theFilter) {
		this->theFilter->startTime = omero::RTimePtr();
		this->theFilter->endTime = omero::RTimePtr();
	    }
	    return this;
	}

	bool ParametersI::isStartTime() {
	    if (0 != this->theFilter) {
		return this->theFilter->startTime != 0;
	    }
	    return false;
	}

	bool ParametersI::isEndTime() {
	    if (0 != this->theFilter) {
		return this->theFilter->endTime != 0;
	    }
	    return false;
	}

	omero::RTimePtr ParametersI::getStartTime() {
	    if (0 != this->theFilter) {
		return this->theFilter->startTime;
	    }
	    return omero::RTimePtr();
	}

	omero::RTimePtr ParametersI::getEndTime() {
	    if (0 != this->theFilter) {
		return this->theFilter->endTime;
	    }
	    return omero::RTimePtr();
	}


	// Parameters.theOptions.leaves, orphan, acquisitionData
	// ===============================================================

	ParametersIPtr ParametersI::leaves() {
	    if (0 == this->theOptions) {
		this->theOptions = new omero::sys::Options();
	    }
	    this->theOptions->leaves = rbool(true);
	    return this;
	}

	ParametersIPtr ParametersI::noLeaves() {
	    if (0 == this->theOptions) {
		this->theOptions = new omero::sys::Options();
	    }
	    this->theOptions->leaves = rbool(false);
	    return this;
	}

	omero::RBoolPtr ParametersI::getLeaves() {
	    if (0 != this->theOptions) {
		return this->theOptions->leaves;
	    }
	    return omero::RBoolPtr();
	}

	ParametersIPtr ParametersI::orphan() {
	    if (0 == this->theOptions) {
		this->theOptions = new omero::sys::Options();
	    }
	    this->theOptions->orphan = rbool(true);
	    return this;
	}

	ParametersIPtr ParametersI::noOrphan() {
	    if (0 == this->theOptions) {
		this->theOptions = new omero::sys::Options();
	    }
	    this->theOptions->orphan = rbool(false);
	    return this;
	}

	omero::RBoolPtr ParametersI::getOrphan() {
	    if (0 != this->theOptions) {
		return this->theOptions->orphan;
	    }
	    return omero::RBoolPtr();
	}

	ParametersIPtr ParametersI::acquisitionData() {
	    if (0 == this->theOptions) {
		this->theOptions = new omero::sys::Options();
	    }
	    this->theOptions->acquisitionData = rbool(true);
	    return this;
	}

	ParametersIPtr ParametersI::noAcquisitionData() {
	    if (0 == this->theOptions) {
		this->theOptions = new omero::sys::Options();
	    }
	    this->theOptions->acquisitionData = rbool(false);
	    return this;
	}

	omero::RBoolPtr ParametersI::getAcquisitionData() {
	    if (0 != this->theOptions) {
		return this->theOptions->acquisitionData;
	    }
	    return omero::RBoolPtr();
	}


	// Parameters.map
	// ===============================================================

	ParametersIPtr ParametersI::add(const std::string& name,
					const omero::RTypePtr& r) {
	    this->map[name] = r;
	    return this;
	}


        ParametersIPtr ParametersI::addId(Ice::Long id) {
	    return add("id", rlong(id));
        }

	ParametersIPtr ParametersI::addId(const omero::RLongPtr& id) {
	    return add("id", id);
	}

	ParametersIPtr ParametersI::addIds(omero::sys::LongList ids) {
	    return addLongs("ids", ids);
	}

	ParametersIPtr ParametersI::addLong(const std::string& name,
					    Ice::Long l) {
	    return add(name, rlong(l));
	}

	ParametersIPtr ParametersI::addLong(const std::string& name,
					    const omero::RLongPtr& l) {
	    return add(name, l);
	}

	ParametersIPtr ParametersI::addLongs(const std::string& name,
					     omero::sys::LongList longs) {
	    omero::RListPtr list = rlist();
	    omero::sys::LongList::const_iterator beg = longs.begin();
	    omero::sys::LongList::const_iterator end = longs.end();
	    while (beg != end) {
		list->add(rlong(*beg));
		beg++;
	    }
	    this->map[name] = list;
	    return this;
	}


    } // sys

} // omero
